sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriter.java

```diff
@@ -56,15 +55,14 @@ public void processElement(ProcessContext ctx) throws Exception {
     checkForFailures();
 
     ProducerRecord<K, V> record = ctx.element();
-    KV<K, V> kv = KV.of(record.key(), record.value());
-
     Long timestampMillis =
```

Comment:
I updated this part - `getPublishTimestampFunction()` has the most priority. If it was not set then we try to get timestamp from `ProducerRecord` (in case if user set it manually before). Otherwise, we just use system time. 
Looks ok for you such approach?

Reply:
Sounds good. We should deprecate `withPublishTimestampFunction()` in favor of writing ProducerRecords.

Reply:
Yes, it's already deprecated